content/courses/native-onchain-development/paging-ordering-filtering-data-frontend.md

```diff
@@ -151,17 +166,26 @@ async function fetchMatchingContactAccounts(
   connection: web3.Connection,
   search: string,
 ): Promise<(web3.AccountInfo<Buffer> | null)[]> {
-  const accounts = await connection.getProgramAccounts(programId, {
-    dataSlice: { offset: 0, length: 0 },
-    filters: [
-      {
-        memcmp: {
-          offset: 13,
-          bytes: bs58.encode(Buffer.from(search)),
-        },
-      },
-    ],
-  });
+  const accounts = (await connection.getProgramAccounts(
+    new PublicKey(MOVIE_REVIEW_PROGRAM_ID),
+    {
+      dataSlice: { offset: 2, length: 18 },
+      filters:
+        search === ""
+          ? []
+          : [
+              {
+                memcmp: {
+                  offset: 6,
+                  bytes: bs58.encode(Buffer.from(search)),
+                },
+              },
+            ],
+    },
+  )) as Array<{
+    pubkey: PublicKey;
+    account: AccountInfo<Buffer>;
+  }>; // Explicitly define the expected structure
```

Comment:
I get that this is being coerced (using `as`) but /why/? What happens if we remove it?

Reply:
> I get that this is being coerced (using `as`) but /why/? What happens if we remove it?

The explicit type definition using as Array<{ pubkey: PublicKey; account: AccountInfo<Buffer>; }> was added because the previous implementation, which didn't have this return type, actually threw an error. By specifying the structure explicitly, it ensures type safety and prevents potential type errors. This also improves the development experience by providing better Intellisense support and avoiding runtime issues.

Reply:
Sure but why does `accounts` need to be coerced? Why is `connection.getProgramAccounts()` not returning the right type without the `as`?